### PR TITLE
[RFC] changes to unique naming in the Module

### PR DESCRIPTION
**Pull request by nickgg**
Summary: This PR aims to fix a few problems with how we do identifier names in the Module and Function:

* The name uniqueness rules tend to append numbers to the end of node names when transforming them, i.e. `relu` -> `relu1`. This is weird when the node name already end with a number, i.e. `relu1` -> `relu11`. It gets weirder when there are similar names, if `relu1` already exists then `relu1` -> `relu12`. We can fix this by inserting a delimiter between the name part and the number part, so e.g. `relu`-> `relu__1` or if taken `relu__2`. If you then transform that node it could become e.g. `relu__3`. This is particularly useful for networks such as ResNet50 which has groups of nodes that share a number since those relationships are preserved through compilation. (btw: Would much prefer another delimiter such as `#` or `|` but we use C-identifier legality and this is pretty much what we've got).
* We ensure uniqueness of Node names within a function, and Storage (Constant/Placeholder) names within a Module but there is no checking across the two name sets. This means that you can't uniquely determine what a given name refers to without additional information (the Kind), since it is valid for it to refer to either a Node or a Storage. There are cases where this is particularly bad, e.g. you could have a Splat named `weights1` which is transformed into a Constant which ends up called `weights1` because no existing Storage has that name, making it unclear a transformation occurred. This PR fixes this by adding a new StringSet to the Module which contains all names used by a Node in any Function. The new rules are:
    - Storage names must be unique across all Storage names in the Module, and all names ever used by a Function owned by that Module.
    - Node names must be unique across all nodes in the Function, and all Storage names that have ever been used in its parent Module. Two different Functions in a Module may have Nodes with the same name, and may refer to the same Storage objects (and thus the same names) but cloned Storage nodes must have a unique name. (I would have liked to just make everything unique, but we rely on cloned functions having the same node names in too many places).
* The most common place where a Storage and Node had the same name was SaveNodes, which basically always create a Placeholder with the same name as the node. I've explicitly changed this so that `createSaveNode(name...)` will create a Placeholder called `name` and a SaveNode called `save_name`. I've fixed all the instances where we look up a Placholder in the bindings with the name of the SaveNode.
* As a side effect, this fixes a bug in the ModelReader where we would modify the names of the models outputs in order to avoid a name collision. E.g. previously Importing then Exporting a model would change the output names to be prepended by "save_", now it doesn't.

Documentation: Comments only?

Test Plan: ran tests under release & asan-debug. Ran tests/images/run.sh

### CODE REVIEW

File: include/glow/Graph/Graph.h [hunk 1]
```diff
@@ -78,15 +81,28 @@ class Module final {
...
   Module() = default;
 
   ~Module();
 
+  /// \returns the prefix part of the provided \p name. E.g. for an input
+  /// of "relu__2" returns "relu".
+  static std::string getPrefix(llvm::StringRef name);
+
   /// \returns unique legal name that's based on the string \p name. Legal
   /// names are legal C identifiers in the form: "[a-zA-Z_][a-zA-Z0-9_]*".
```
[the remark this answers is not in the document]
Comment (nickgg) [reply include/glow/Graph/Graph.h]: Sorry what do you mean, update the comment or regenerate the doxygen docs?

**Comment by nickgg** [comment]
Should have fixed the example/bundle issues.

File: include/glow/Graph/Graph.h [hunk 2]
```diff
@@ -64,7 +64,10 @@ class Module final {
   TypesList types_{};
   /// Stores a list of unique variable names that were used by the module at
```
[the remark this answers is not in the document]
Comment (nickgg) [reply include/glow/Graph/Graph.h]: @jfix71 I fixed that one, and a few more uses of "Variable" but there are a bunch more in Graph.h remaining.

**Comment by ayermolo** [comment]
@nickgg 
You changed SaveNode *Function::createSave(llvm::StringRef name, NodeValue input) to add "_save", but not 
SaveNode *Function::createSave(llvm::StringRef name, NodeValue input,
                               Placeholder *output)

Is that an oversight?

**Pull request closed without merging**